File: problem-solving/lc-100-easy-same-tree/1-recursive.go
/*
https://leetcode.com/problems/same-tree/

Complexity
----------
Time and Space: O(N)

Note
----
Iterative approach is not very readable for this question,
recursive is preferrable due to clean and short code.
*/

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSameTree(p *TreeNode, q *TreeNode) bool {
    if p == nil || q == nil {
        return p == q
    }
    if p.Val != q.Val {
        return false
    }
    return isSameTree(p.Left, q.Left) && isSameTree(p.Right, q.Right)
}

File: problem-solving/lc-98-medium-validate-binary-search-tree/1-recursive.py
"""
https://leetcode.com/problems/validate-binary-search-tree/

Complexity
----------
Time: O(N)
Space: O(N)
"""

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def validate(root, mn, mx):
            if root is None:
                return True
            if root.val <= mn or root.val >= mx:
                return False
            return (validate(root.left, mn, root.val) and
                    validate(root.right, root.val, mx))

        return validate(root, -math.inf, math.inf)
